bpf/bpf_lxc.c

```diff
@@ -696,47 +758,50 @@ TAIL_CT_LOOKUP6(CILIUM_CALL_IPV6_CT_EGRESS, tail_ipv6_ct_egress, CT_EGRESS,
 		is_defined(ENABLE_PER_PACKET_LB),
 		CILIUM_CALL_IPV6_FROM_LXC_CONT, tail_handle_ipv6_cont)
 
-static __always_inline int __tail_handle_ipv6(struct __ctx_buff *ctx,
-					      __s8 *ext_err __maybe_unused)
+static __always_inline int __tail_handle_ipv6(struct __ctx_buff *ctx)
 {
 	void *data, *data_end;
 	struct ipv6hdr *ip6;
-	int ret;
+	__s8 ext_err = 0;
+	int ret = 0;
 
 	if (!revalidate_data_pull(ctx, &data, &data_end, &ip6))
-		return DROP_INVALID;
+		return send_drop_notify_error_ext(ctx, SECLABEL, DROP_INVALID, ext_err,
+						  CTX_ACT_DROP, METRIC_EGRESS);
 
 	/* Handle special ICMPv6 NDP messages, and all remaining packets
 	 * are subjected to forwarding into the container.
 	 */
 	if (unlikely(is_icmp6_ndp(ctx, ip6, ETH_HLEN))) {
 		ret = icmp6_ndp_handle(ctx, ETH_HLEN, METRIC_EGRESS);
 		if (IS_ERR(ret))
-			return ret;
+			return send_drop_notify_error_ext(ctx, SECLABEL, ret, ext_err,
+							CTX_ACT_DROP, METRIC_EGRESS);
 	}
 
 	if (unlikely(!is_valid_lxc_src_ip(ip6)))
-		return DROP_INVALID_SIP;
+		return send_drop_notify_error_ext(ctx, SECLABEL, DROP_INVALID_SIP, ext_err,
+						  CTX_ACT_DROP, METRIC_EGRESS);
 
 #ifdef ENABLE_PER_PACKET_LB
 	/* will tailcall internally or return error */
-	return __per_packet_lb_svc_xlate_6(ctx, ip6, ext_err);
+	/* TODO this should report its own drops. */
+	ret = __per_packet_lb_svc_xlate_6(ctx, ip6, &ext_err);
+	if (IS_ERR(ret))
+		return send_drop_notify_error_ext(ctx, SECLABEL, ret, ext_err,
+						  CTX_ACT_DROP, METRIC_EGRESS);
 #else
 	/* won't be a tailcall, see TAIL_CT_LOOKUP6 */
 	return tail_ipv6_ct_egress(ctx);
 #endif /* ENABLE_PER_PACKET_LB */
+
+	return ret;
 }
 
 __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV6_FROM_LXC)
 int tail_handle_ipv6(struct __ctx_buff *ctx)
 {
-	__s8 ext_err = 0;
-	int ret = __tail_handle_ipv6(ctx, &ext_err);
-
-	if (IS_ERR(ret))
-		return send_drop_notify_error_ext(ctx, SECLABEL, ret, ext_err,
-						  CTX_ACT_DROP, METRIC_EGRESS);
```

Comment:
It is less verbose, for sure - I was trying to consistently stick to an `always report at source` principle, since not following that and making assumptions in a generic macro func about what specific calling funcs will do is how we got this problem in the first place.

Nested macro option you mentioned is a reasonable compromise tho.

Reply:
Reverted, I don't feel strongly about it and it's tangental.